stdenv.mkDerivation: overlay style overridable recursive attributes

roberth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

 - Allows to get rid of confusing `rec { }` in `mkDerivation` calls. Overriding an attribute in this style causes derived attributes to be overridden as expected.

 - Allows `passthru` attributes to reference the outputs

 - Fixes #119407, the problem where if you use `overrideAttrs`, the tests in `passthru` still use the old package.

 - Should allow cleanup of python and rust packaging functions by using "overlay" layers on a prototype derivation, rather than the current workarounds.

See updated manual for usage explanation.

Update: this problem was also brought up in this discussion: https://github.com/NixOS/nixpkgs/pull/119731#pullrequestreview-638309038

Update: "overlay" part minus `public` was also discovered and implemented by lilyball https://github.com/NixOS/nixpkgs/pull/94198

<!--
# :rotating_light: Please prioritize reviews and fixes for the release first!

### Consider that your reviewer is already busy with stabilization work and other reviews, so now is not really a time for refactoring, unless you need the change to solve a significant problem. Thank you.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing **not applicable:** changes are evaluation-only ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

roberth committed:
hello: Define a passthru test via new mkDerivation self arg

roberth committed:
stdenv.mkDerivation: Allow overriding of recursive definitions

See updated manual for further explanation.

davidak commented on pkgs/applications/misc/hello/default.nix:
```diff
@@ -1,11 +1,11 @@
+{ lib, runCommand, stdenv, fetchurl }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation ({ this, ... }: {
```
i don't like the overhead this adds. the derivation gets a lot less readable

roberth replied:
I've reintroduced the `rec` keyword, so the changes are minimal. It wasn't unreadable before and now it's barely changed at all.

Now `version` is in scope again, which, if you override it won't affect `src`, as usual. That's the way derivations are currently written. In my opinion, `this.version` has better behavior, you don't have to agree and I'm ok to stick to the old format until the community decides otherwise. I'm only interested in making `passthru` behave correctly. The ability to use `src = f this.version;` is just side-catch.

davidak commented on pkgs/applications/misc/hello/default.nix:
```diff
@@ -18,9 +18,16 @@ stdenv.mkDerivation rec {
   };
-}
+
+  passthru.tests.run = runCommand "hello-test-run" {
```
tests should be added in a separate file
maybe then this whole concept does not work anymore

roberth replied:
This works just fine.

roberth committed:
hello: Use callPackage for test

davidak commented on pkgs/applications/misc/hello/default.nix:
```diff
@@ -23,4 +23,6 @@ stdenv.mkDerivation rec {
   };
-}
+
+  passthru.tests.run = callPackage ./test.nix { hello = this; };
```
i don't really understand why this is an improvement.
wouldn't it have the same behavior when you remove `hello = this;`?
like in this example https://github.com/NixOS/nixpkgs/blob/d600f006643e074c2ef1d72e462e218b647a096c/pkgs/tools/misc/phoronix-test-suite/default.nix#L23

roberth replied:
That way you'll be testing the un-overridden version if you use `overrideAttrs`.

```
nix-instantiate --expr '((import ./. {}).phoronix-test-suite.overrideAttrs (o: { dontBuild = true; })).tests.simple-execution' '(import ./. {}).phoronix-test-suite.tests.simple-execution'
/nix/store/aj6lv43p8hlkhlbd6klljgami4cb48qv-phoronix-test-suite-tests.drv
/nix/store/aj6lv43p8hlkhlbd6klljgami4cb48qv-phoronix-test-suite-tests.drv
```

So, same test although the first one had `overrideAttrs` applied to it. Should have been different.

davidak replied:
i see. thanks

davidak commented:
I see how this can be useful in the case of using `overrideAttrs`, but i don't like that we would have to add this overhead to every package.

It is a common use case to use the current package in `passthru.tests`, so could we achieve this with a more elegant implementation? I don't have to manually specify something to be able to use `$out` in buildPhase, why do i have to for `passthru.tests`?

When thinking this further, what if i override a package and also test another package that depend on it in it's `passthru.tests`?

roberth commented:
> what if i override a package and also test another package that depend on it in it's `passthru.tests`?

If you don't use `this` in a `tests` attribute, you're not testing the final package. In such a case you should override that package to use `this`. You could use `pkgs.extend` for this purpose.

> I don't have to manually specify something to be able to use $out in buildPhase, why do i have to for passthru.tests?

`$out` is a shell variable that only exists inside the derivation. When you write a test in a separate derivation, its `$out` will not equal the `$out` of the original derivation, so you'll need to use the Nix expression language to be able to reference it. Variables in the Nix expression are mostly lexically scoped and immutable, unlike bash's global-by-default mutable scope. The only exception is the `with` keyword, which is only a fallback for variable references that would be unresolvable otherwise.

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -18,7 +18,16 @@ in rec {
+      this = mkDerivation_ (transform: mkDerivation (transform fn)) (fn { inherit this; });
+    in this;
+
+  mkDerivation_ = remakeDerivation:
```
I would make this take `overrideAttrs` not `remakeDerivation`, so we can do all the "extensions and fixed points" combinatorics above.

roberth replied:
I've made it look like `makeExtensible` instead and documented the differences. That also means I'm adding `overrideAttrs` in `makeDerivationExtensible`, just after applying the user's function.
So this is closer to the `lib` functions and moves all of the overriding stuff outside `mkDerivation_`.

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -354,7 +363,14 @@ in rec {
+                  oldf:
+                    # this function body is exactly like a mkDerivation argument
+                    { this, ... }@t:
+                    let attrs = oldf t; in attrs // f (attrs // { inherit this; })
```
f's parameter is like `super`, but what if we did another round of `builtins.isFunction` trickery so we could have `super` and `self` parameters? Then this would better match other such things in Nixpkgs, and also, we can get rid off `rec { ... }` `mkDerivation` arguments which are hard to override.

I think we can also use `lib.fixed-points.extend` at that point.

roberth replied:
:100: 

I thought of doing `self: super:` but considered it too ugly. I didn't quite realize we could do reflection inside the partially applied function cheaply.

> we can get rid off rec { ... } mkDerivation arguments which are hard to override.

This is definitely a possibility, even before the refactor, but I got an objection about this (big) change earlier. Whether or not the community wants to change "style" is not in any way dictated by this PR though.

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -18,7 +18,16 @@ in rec {
+        if builtins.isFunction fnOrAttrs
+        then fnOrAttrs
+        else _: fnOrAttrs;
+      this = mkDerivation_ (transform: mkDerivation (transform fn)) (fn { inherit this; });
```
Oh, I see, it's just `this` being passed the first time. So we are all `this` / `super` today, with no `self`. Hmm.

roberth replied:
Sounds about right. This context is outdated though, so I'll resolve this conversation.

Ericson2314 commented:
Also, I would check out the saga of https://github.com/NixOS/nixpkgs/pull/27319 PR, which had similar goals long before, but ended up being reverted for perf reasons (IIRC).

roberth commented:
I've updated the docs. It now includes this example which I think is particularly nice:

```
    passthru.appendPackages = packages':
      self.overrideAttrs (newSelf: super: {
        packages = super.packages ++ packages';
      });
```


> Also, I would check out the saga of #27319 PR, which had similar goals long before, but ended up being reverted for perf reasons (IIRC).

This is quite different and I've checked performance with `bench` aka `criterion` on an idle machine.

```
bench -L 1800 'nix-instantiate -A hello -A pandoc -A cassandra -A rabbitmq-server'
```

master:

```
benchmarking nix-instantiate -A hello -A pandoc -A cassandra -A rabbitmq-server
time                 2.206 s    (2.205 s .. 2.207 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.205 s    (2.204 s .. 2.206 s)
std dev              2.875 ms   (1.878 ms .. 4.684 ms)
```

overridable:

```
benchmarking nix-instantiate -A hello -A pandoc -A cassandra -A rabbitmq-server
time                 2.216 s    (2.215 s .. 2.217 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.216 s    (2.215 s .. 2.216 s)
std dev              1.822 ms   (1.462 ms .. 2.350 ms)
```

So the slowdown is only 0.45%, which seems like a fair price to pay for correctness.

Memory usage has increased slightly as well.

```diff
--- master.stats        2021-05-02 23:46:59.409895694 +0200
+++ overridable.stats   2021-05-02 23:47:12.699873034 +0200
@@ -1,27 +1,27 @@
 {
-  "cpuTime": 1.72167,
+  "cpuTime": 1.70415, # noisy; ignore; see proper stats above
   "envs": {
-    "number": 565107,
-    "elements": 1036066,
-    "bytes": 17330240
+    "number": 587547,
+    "elements": 1063062,
+    "bytes": 17905248
   },
   "list": {
-    "elements": 364943,
-    "bytes": 2919544,
+    "elements": 373857,
+    "bytes": 2990856,
     "concats": 53359
   },
   "values": {
-    "number": 2043552,
-    "bytes": 49045248
+    "number": 2058343,
+    "bytes": 49400232
   },
   "symbols": {
-    "number": 92751,
-    "bytes": 2812685
+    "number": 92755,
+    "bytes": 2812733
   },
   "sets": {
-    "number": 189351,
-    "bytes": 175026936,
-    "elements": 7229672
+    "number": 198270,
+    "bytes": 177275760,
+    "elements": 7320400
   },
   "sizes": {
     "Env": 16,
@@ -29,15 +29,15 @@
     "Bindings": 8,
     "Attr": 24
   },
-  "nrOpUpdates": 105296,
-  "nrOpUpdateValuesCopied": 6123991,
-  "nrThunks": 1545207,
-  "nrAvoided": 963446,
-  "nrLookups": 596064,
-  "nrPrimOpCalls": 382588,
-  "nrFunctionCalls": 501801,
+  "nrOpUpdates": 109755,
+  "nrOpUpdateValuesCopied": 6205698,
+  "nrThunks": 1559996,
+  "nrAvoided": 989027,
+  "nrLookups": 601181,
+  "nrPrimOpCalls": 387046,
+  "nrFunctionCalls": 517808,
   "gc": {
     "heapSize": 402915328,
-    "totalBytes": 287925024
+    "totalBytes": 292058624
   }
 }
```

roberth commented:
Updated stats after optimization:

```diff
-    "totalBytes": 287925024
+    "totalBytes": 288061888
```

Full stats

```diff
--- master.stats        2021-05-03 01:06:09.010757018 +0200
+++ overridable-positional-removedAttrs-extends-specialized.stats       2021-05-03 01:26:35.772202431 +0200
@@ -5,29 +5,29 @@
 /nix/store/ajsbx332p1rylbysvas25p6wasy86g4z-cassandra-3.11.10.drv
 /nix/store/n7y0pnm6kkdvx48p92dadl6f613zxzqg-rabbitmq-server-3.8.9.drv
 {
-  "cpuTime": 1.67284,
+  "cpuTime": 1.71675,  # noisy; ignore. Will post new benchmark.
   "envs": {
-    "number": 565107,
-    "elements": 1036066,
-    "bytes": 17330240
+    "number": 581776,
+    "elements": 1052930,
+    "bytes": 17731856
   },
   "list": {
-    "elements": 364943,
-    "bytes": 2919544,
+    "elements": 320901,
+    "bytes": 2567208,
     "concats": 53359
   },
   "values": {
-    "number": 2043552,
-    "bytes": 49045248
+    "number": 2041929,
+    "bytes": 49006296
   },
   "symbols": {
-    "number": 92751,
-    "bytes": 2812685
+    "number": 92757,
+    "bytes": 2812770
   },
   "sets": {
-    "number": 189351,
-    "bytes": 175026936,
-    "elements": 7229672
+    "number": 189356,
+    "bytes": 175032496,
+    "elements": 7229902
   },
   "sizes": {
     "Env": 16,
@@ -35,15 +35,15 @@
     "Bindings": 8,
     "Attr": 24
   },
-  "nrOpUpdates": 105296,
-  "nrOpUpdateValuesCopied": 6123991,
-  "nrThunks": 1545207,
-  "nrAvoided": 963446,
-  "nrLookups": 596064,
-  "nrPrimOpCalls": 382588,
-  "nrFunctionCalls": 501801,
+  "nrOpUpdates": 105298,
+  "nrOpUpdateValuesCopied": 6124018,
+  "nrThunks": 1543582,
+  "nrAvoided": 942353,
+  "nrLookups": 600523,
+  "nrPrimOpCalls": 387046,
+  "nrFunctionCalls": 516493,
   "gc": {
     "heapSize": 402915328,
-    "totalBytes": 287925024
+    "totalBytes": 288061888
   }
 }
```

roberth commented:
And new timing stats: 0.18% increase.

master:

```
benchmarking nix-instantiate -A hello -A pandoc -A cassandra -A rabbitmq-server
time                 2.208 s    (2.207 s .. 2.209 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.206 s    (2.205 s .. 2.207 s)
std dev              3.017 ms   (2.088 ms .. 5.007 ms)
```

overridable with optimizations:

```
benchmarking nix-instantiate -A hello -A pandoc -A cassandra -A rabbitmq-server
time                 2.212 s    (2.211 s .. 2.213 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.211 s    (2.211 s .. 2.212 s)
std dev              2.119 ms   (1.562 ms .. 3.367 ms)
```

danieldk commented on doc/stdenv/meta.chapter.md:
```diff
@@ -148,6 +148,36 @@ The NixOS tests are available as `nixosTests` in parameters of derivations. For
+Alternatively, you can specify other derivations as tests. You can make use of
+the optional parameter (here: `self`) to inject the correct package without
+relying on non-local definitions, even in the presence of `overrideAttrs`. This
+means `(mypkg.overrideAttrs f).tests` will be as expected, as long as the
```
```suggestion
means `(mypkg.overrideAttrs f).passthru.tests` will be as expected, as long as the
```

roberth replied:
Thanks! These are the same attributes, but your change makes the text easier to understand.

danieldk replied:
Ah, I didn't even realize that :) .

roberth committed:
doc/stdenv/meta: tests -> passthru.tests

Make the text understandable without knowing that these are identical.

Co-authored-by: Daniël de Kok <me@github.danieldk.eu>

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -19,6 +19,58 @@ in rec {
+              let x = f0 super;
+              in
+                if builtins.isFunction x
+                then f0 self super # double call looks inefficient, but `f0 super` was a cheap thunk
```
Also maybe point out since `self` is the first arg currying coudn't help anyways.

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -19,6 +19,58 @@ in rec {
+                then f0 self super # double call looks inefficient, but `f0 super` was a cheap thunk
+                else x;
+          in
+            makeDerivationExtensible (self: let super = rattrs self; in super // f self super)) (rattrs r);
```
I would maybe indent this differently so `(f0: ...)` and `(rattrs r)` have the same indent?

roberth replied:
I've `(rattrs r)` on its own line.

Ericson2314 replied:
err I mean the previous arg has the extra paren at the end: these are arguments to `mkDerivation_` not `makeDerivationExtensible`.

roberth replied:
:facepalm:

That was very necessary. Thanks :)

danieldk reviewed (approved):
This is really a nice extension.  I tried to check the details and LGTM. Should be taken with a grain of salt though, since I am not extremely familiar with this part of `mkDerivation`.

roberth committed:
stdenv.mkDerivation: Make self more overlay-like; use self.public

`self` is now arguments, like `super`. The final package is in
`self.public`.

roberth commented:
I've changed self to be more overlay-like. The final package is in `self.public`.

This is implemented in 45503359684baa1efd5d145a29e35c431f9f7b74 (stdenv.mkDerivation: Make self more overlay-like; use self.public). Performance should be similar, because the changes are minimal: 1 extra let binding. The extra `//` isn't accessed in current use of `mkDerivation`.

roberth commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -9,8 +9,65 @@ let
+
 in
 
+# TODO(@roberth): inline makeOverlayable; reindenting whole rest of this file.
+makeOverlayable (overrideAttrs:
```
Not performing this refactoring/reformatting in this PR to keep the diff readable.

andir replied:
Why is it better to have it inlined? I kinda like having it up there with a name (that tells me what it does).

roberth replied:
I've been paranoid about performance while writing this. Removing.

roberth commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -9,8 +9,65 @@ let
+      else makeDerivationExtensible0 mkDerivationSimple fnOrAttrs;
+
+  # Based off lib.makeExtensible, with modifications:
+  makeDerivationExtensible = mkDerivationSimple: rattrs:
```
The `mkDerivationSimple` parameter will become a let binding reference, but that (simple) refactoring creates a huge reindentation diff, so I'll do it afterwards.

roberth replied:
Not worth the hassle and loss of git blame.

Artturin replied:
> loss of git blame.
fyi we can now add revs to `.git-blame-ignore-revs` to ignore them in blame

roberth reviewed (commented):
Rebased; reviews are very much appreciated.

roberth commented:
@Ericson2314 can we move forward with this?

andir commented on doc/stdenv/meta.chapter.md:
````diff
@@ -175,6 +175,37 @@ The NixOS tests are available as `nixosTests` in parameters of derivations. For
+
+```nix
+# my-package/default.nix
+{ stdenv, callPackage, ... }:
````
I think you should drop the `...` here and in the next file. It doesn't matter and has no usage. It would just confuse newcomers.

roberth replied:
Done.

andir commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -9,8 +9,65 @@ let
+    in public;
+
+  # Exactly makeDerivationExtensible (_: attrs), but specialized for performance
+  makeDerivationExtensible0 = mkDerivationSimple: attrs:
```
Perhaps that name could be better in this case? What meaning does that `0` even have? It is yet another version of the function, but we aren't writing math here.

roberth replied:
Is `makeDerivationExtensibleConst` better?

It kind of is math, tbh, but that's no reason to adopt a worse syntax. Updating the comment though to use an equality operator rather than "exactly".

andir reviewed (dismissed):
I've added a couple of remarks. Nothing major but in general I am all for this change.

Good work 👍🏻

roberth committed:
make-derivation.nix: Remove unnecessary TODO

The goal was to keep the number of calls to a minimum, but that
would come at the cost of clobbering git blame and possibly not
improving readability.

roberth committed:
Use finalAttrs instead of self for mkDerivation "overlay"

roberth committed:
doc/using/overrides: Update for overlay style mkDerivation overrideAttrs

jtojnar commented:
Looks like this is actually a duplicate of https://github.com/NixOS/nixpkgs/pull/94198. I completely forgot about that PR.

roberth commented:
> Looks like this is actually a duplicate of #94198. I completely forgot about that PR.

Right, plus adding a `public` attribute to `self` to represent the end result, adding documentation, minus some changes to packages, which could be cherry-picked.

lilyball commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -305,6 +305,61 @@ The script will be usually run from the root of the Nixpkgs repository but you s
+Instead, the definition references `finalAttrs`, allowing users to change `withFeature`
+consistently with `overrideAttrs`.
+
+`finalAttrs` also contains the attribute `public`, which represents the final package,
+including the output paths, etc.
```
What happens if the derivation attributes included a field called `public`?

What happens if I write something like `drv.overrideAttrs (final: final)`? Won't that turn the package into an input of itself?

roberth replied:
The `mkDerivation` caller or overrider can still set `public`, but the overrider can only read that value back from `previousAttrs`; not `finalAttrs`. `previousAttrs` is also the legacy one, so this won't break derivations that already have a `public` attr, which would be out-of-tree ones.

> `drv.overrideAttrs (final: final)`

This parameter shouldn't be called `final`, as a one-argument `overrideAttrs` behaves as it always did: `previousAttrs` (or `old`). It is equal to `drv`.

> something like `drv.overrideAttrs (final: final)`

Just like with overlays, `drv.overrideAttrs (finalAttrs: previousAttrs: finalAttrs)` produces a detectable infinite recursion.

I've managed to insert a custom error message, even if it's a bit hacky. Here's a somewhat realistic attempt of trying to define an attr only when another attr is defined. This is fundamentally impossible in Nix (as it currently exists; I believe @Infinisil tried to change this once. I don't know if that project has succeeded.), because the attributes part of the attrset is a strict data structure. 

```
nix-repl> (stdenv.mkDerivation { name = "hi"; public = "hi"; }).overrideAttrs (finalAttrs: previousAttrs: lib.optionalAttrs (finalAttrs?foo) { bar = "Only if we'll have foo for sure!"; }) 
error: infinite recursion encountered

       at /home/user/h/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:27:22:

           26|       # An infinite recursion here can be caused by having the attribute names of expression `e` in .overrideAttrs(finalAttrs: previousAttrs: e) depend on `finalAttrs`. Only the attribute values of `e` can depend on `finalAttrs`.
           27|       args = rattrs (args // { inherit public; });
             |                      ^
           28|       #              ^^^^
```

roberth replied:
Infinite recursion involving `builtins.derivation`. I feel like this is more unlikely and it prints a somewhat informative error message:

```
$ nix-instantiate --expr 'with import ./. {}; (stdenv.mkDerivation { name = "hi"; public = "hi"; }).overrideAttrs (finalAttrs: previousAttrs: { bar = finalAttrs.public; })' --show-trace
error: infinite recursion encountered

       at //builtin/derivation.nix:19:19:
       … while evaluating the attribute 'bar' of the derivation 'hi'

       at /home/user/h/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:270:7:

          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let

```

roberth replied:
A new way of self-bootstrapping. Wouldn't recommend because this doesn't have cross compilation splicing support, but it's interesting. Maybe cross compilation splicing could be done via a function similar to `overrideAttrs`, invoked by the cross compilation code. I've played around with the idea of moving the code that improves the output attrs into a postcomposed function that would defined in the fixpoint to be overridable. Perhaps that would allow it, but the interaction between the `overrideAttrs` and such a function isn't easy to get right.

Anyway, for the curious:

```
stdenv.mkDerivation(finalAttrs: {
  name = "bootstrapper";
  nativeBuildInputs = [ finalAttrs.bootPkg ];
  bootPkg = finalAttrs.public.overrideAttrs (_: { bootPkg = null; });
  dontUnpack = true;
  installPhase = ''
    mkdir -p $out/bin
    if test -n "''${bootPkg}"; then
      $bootPkg/bin/boot "world" >$out/msg
    fi
    dollar='$'
    cat >$out/bin/boot <<EOF
    #!${runtimeShell}
    echo "Hello ''${dollar}1!"
    if test -e $out/msg; then
      echo "also"
      cat $out/msg
    fi
    EOF
    chmod a+x $out/bin/boot
  '';
})
```

```
$ ./result/bin/boot there
Hello there!
also
Hello world!
```

lilyball replied:
Sorry yes I did mean `drv.overrideAttrs (old: old)`. I wasn't reading carefully enough though to realize where the `public` attr went.

In any case, what happens if a derivation already has a `public` attr? Does `finalAttrs.public` just hide that original `public` attr? Can you say `finalAttrs.public.public` to get at it?

roberth replied:
> In any case, what happens if a derivation already has a `public` attr? Does `finalAttrs.public` just hide that original `public` attr?

Yes

> Can you say `finalAttrs.public.public` to get at it?

Yes, `mkDerivation` returns most if not all of its arguments, so you can.

```
nix-repl> (stdenv.mkDerivation (finalAttrs: { public = true; })).public                          
true

nix-repl> (stdenv.mkDerivation (finalAttrs: { public = true; viaFinal = finalAttrs.public.public; })).viaFinal
true

nix-repl> p = stdenv.mkDerivation (finalAttrs: { public = true; })

nix-repl> (p.overrideAttrs (previousAttrs: { via = previousAttrs.public; })).via
true

nix-repl> (p.overrideAttrs (finalAttrs: previousAttrs: { via = finalAttrs.public.public; })).via
true
```

lilyball replied:
Ok thank you, that satisfies my questions.

roberth committed:
mkDerivation: Add error hint for infinite recursion

roberth commented:
> Are subsequent changes needed for buildPythonApplication, buildRustApplication and so on to support this?

It is compatible, so no change is necessary, but these functions may take advantage of it.

It seems that wiring around `mk-python-derivation.nix` could be simplified by reframing it as a "prototype" derivation (as in prototype inheritance), which is extended to become a specific python derivation. Perhaps `buildPythonApplication` could be backed by a prototype that is derived from the basic python prototype. I haven't tried this; it's probably not that simple, but if it is, it could be a significant simplification of the glue.

I can imagine that, to take full advantage of this, we'll want to add a new `mkDerivation` parameter to filter away some "local" or "intermediate" parameters that should not or can not be passed to `builtins.derivationStrict`. That is not the only solution though and it isn't necessary for this PR, so let's consider it in a follow-up PR.

roberth commented:
Any objections to a merge 2 weeks from now? Can be technical, process, anything.

Synthetica9 commented on pkgs/applications/misc/hello/default.nix:
```diff
@@ -1,16 +1,17 @@
 
   src = fetchurl {
-    url = "mirror://gnu/hello/${pname}-${version}.tar.gz";
+    url = "mirror://gnu/hello/${finalAttrs.pname}-${finalAttrs.version}.tar.gz";
```
```suggestion
    url = "mirror://gnu/hello/${finalAttrs.public.name}.tar.gz";
```

lilyball replied:
I disagree with this suggested change. If I'm overriding the package I should have some means of giving it a different name (so I can visually distinguish it) without breaking the `src` attribute. With the code as written, I can provide a `name` attr and leave `pname` alone. This is admittedly slightly odd, but it would work. Alternatively the package could just hardcode the name `hello` in the URL, like it does in the directory name, and that would leave both `name` and `pname` free for use.

roberth replied:
I've hardcoded the use of `pname` in the url, because it doesn't make sense to DRY things we don't control and are just a given.

Synthetica9 commented:
I have thought about this some more, and while I think this is really cool, I'm a bit worried about "secondary builders" like `buildPythonPackage` and so many more. How easy is it to adapt those builders to this new model?

FRidh commented:
While not continued we spoke in an RFC about overriding layers. The idea of a prototype derivation to me seems to go in a different direction. https://github.com/NixOS/rfcs/pull/67#issuecomment-729051585

roberth commented:
> The idea of a prototype derivation to me seems to go in a different direction. [NixOS/rfcs#67 (comment)](https://github.com/NixOS/rfcs/pull/67#issuecomment-729051585)

That direction can still be taken. An extra parameter doesn't affect that idea or its feasibility.

roberth committed:
rl-2205: Add entry for overlay-style mkDerivation overriding

lilyball commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -9,8 +9,72 @@ let
+
+  makeOverlayable = mkDerivationSimple:
+    fnOrAttrs:
+      if builtins.isFunction fnOrAttrs
```
Perhaps `lib.isFunction` instead? This way functors will work.

roberth replied:
I've considered it, but
 - this is performance sensitive code
 - functors are for "objects" that double as a function, usually for compatibility reasons. I don't expect these to be used here
 - it is more of a breaking change, as someone may already be using this to turn a package into a functor, in which case the functor _must not_ be invoked
 - it doesn't add new functionality. An eta expansion of the functor is enough to make it work: `mkDerivation (finalAttrs: functorObj finalAttrs)`

lilyball commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -9,8 +9,72 @@ let
+                # arise from flipping an overlay's parameters in some cases.
+                let x = f0 super;
+                in
+                  if builtins.isFunction x
```
This should probably be `lib.isFunction` too, so functors work.

roberth replied:
Same as https://github.com/NixOS/nixpkgs/pull/119942#discussion_r802573480

lilyball commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -9,8 +9,72 @@ let
+          f = self: super:
+            let x = f0 super;
+            in
+              if builtins.isFunction x
```
`lib.isFunction` here too

roberth replied:
Same as https://github.com/NixOS/nixpkgs/pull/119942#discussion_r802573480

virusdave commented on doc/stdenv/stdenv.chapter.md:
````diff
@@ -317,6 +317,61 @@ The script will be usually run from the root of the Nixpkgs repository but you s
+If you pass a function to `mkDerivation`, it will receive as its argument the final arguments, considering use of `overrideAttrs`. For example:
+
+```nix
+mkDerivation (finalAttrs: {
+  pname = "hello";
+  withFeature = true;
+  configureFlags =
+    lib.optionals finalAttrs.withFeature ["--with-feature"];
+})
````
> and in the future hopefully we can slap lib.makeOverridable in the fetchers so I can override their args instead of the resulting derivation

See relevant https://github.com/NixOS/nixpkgs/pull/158018

roberth replied:
Let's scope this out and follow it up in #158018.

roberth committed:
stdenv.md: Clarify overrideAttrs sentence

I weirded my English.

roberth commented on doc/stdenv/stdenv.chapter.md:
````diff
@@ -317,6 +317,61 @@ The script will be usually run from the root of the Nixpkgs repository but you s
+If you pass a function to `mkDerivation`, it will receive as its argument the final arguments, considering use of `overrideAttrs`. For example:
+
+```nix
+mkDerivation (finalAttrs: {
````
`finalAttrs` is unambiguous, avoiding name shadowing when used in an overlay that uses names conforming to what's expected from flakes - `final` and `prev` are taken.

`self` is from the older overlay naming convention, which is not great but less of an issue for name shadowing reasons. I would like to include `attrs` though, because it conforms with `overrideAttrs`, hinting a tiny bit that it's about the `mkDerivation` arguments, not the result.

roberth commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -317,6 +317,61 @@ The script will be usually run from the root of the Nixpkgs repository but you s
+Instead, the definition references `finalAttrs`, allowing users to change `withFeature`
+consistently with `overrideAttrs`.
+
+`finalAttrs` also contains the attribute `public`, which represents the final package,
+including the output paths, etc.
```
I imagined this to be _like_ using the (currently non-existent) `mkPackage` function I wrote about here https://github.com/NixOS/rfcs/pull/92#discussion_r671741003 for working with packages that use delayed IFD and hence contain at least two derivations.

Most of the `mkPackage` wiring can be implemented by a generic "encapsulation" function that consists of an overlay structure where the direct overlay attrs are protected, by returning only the `public` attr, and `super` contains an overriding function (like `overrideAttrs`).

Perhaps I should implement that function? It makes little sense to conform to something that barely exists yet.

roberth committed:
hello: Make pname overridable without breaking src.url

roberth commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -317,6 +317,61 @@ The script will be usually run from the root of the Nixpkgs repository but you s
 
+### Recursive attributes in `mkDerivation`
+
+If you pass a function to `mkDerivation`, it will receive as its argument the final arguments, considering use of `overrideAttrs`. For example:
```
Changed. I've made it a bit more detailed than that, to dispel some of the magic.

edolstra commented:
I'm not in favor of adding more machinery to `stdenv.mkDerivation` and `.override*`, which are already way too complicated and ad hoc. This does not improve discoverability or usability of Nixpkgs (e.g. it doesn't help users discover what can be overriden in a package, and it doesn't help them decide whether to use `.override`, `.overrideAttrs`, `.public.override*` etc). I'm really only on board with switching to something like the NixOS module system (e.g. [Nix modules](https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d) or the Nickel contracts), because that's discoverable and has a consistent override mechanism.

In any case, since this PR proposes a pretty big change in how we write packages, I think it should be discussed in an RFC.

roberth commented:
@edolstra I think you're overestimating the complexity of this PR. It just adds a parameter, in a backwards compatible manner.
The implementation looks complicated, but that's only because I've inlined some code for performance.
The overriding logic and the derivation logic can be split, but I haven't done so (yet?) to avoid touching more lines than necessary.

> it doesn't help them decide whether to use .override, .overrideAttrs, .public.override* etc

There's no such thing as `.public.override`.  `public` is just the return value of `mkDerivation`. `public` is only relevant for _defining_ `passthru` packages anyway. It does not exist outside of `finalAttrs`.

I don't see how any code could help someone decide which function's arguments to override. What this code does do, is make the overriding mechanism more powerful, so new code can do more without having to introduce more overridable functions. It allows the number of choices to be kept low.

> This does not improve discoverability

Sure that would be nice, but it is in no way a requirement for this PR to be an improvement. We can all think of many problems this PR does not solve and that's fine, because the future doesn't end after this PR.

>  or usability of Nixpkgs

It does allow solving the usability problem that is `rec` in `mkDerivation`, which is not an overriding mechanism despite the presence of an overriding mechanism, which is confusing.

> I'm really only on board with switching to something 

I'm also on board with those ideas, but those are rewrites instead of incremental improvements that have a positive effect today.

> like the NixOS module system

Haskell.nix does this, only for Haskell packages and it is slow.

> [or] (e.g. [Nix modules](https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d)

Those are not available and they don't solve the problem of needing overrides before/after each function that transforms the package attrs. If they do, that amounts to a Nixpkgs rewrite.

> or the Nickel contracts

Similar to Nix modules, but with the added cost of migrating to a different language, or making it compatible with Nix. Last time I checked, Nix compatibility wasn't an explicit goal.

> because that's discoverable

That's neat, but not a requirement for this incremental improvement.

> and has a consistent override mechanism.

As does this.

> since this PR proposes a pretty big change in how we write packages

It's a _small_ change in how we _can choose_ to write packages.

I don't look forward to working on an RFC that seems unlikely to bring substantial improvements to a PR you disagree with. I hope I'm seeing this wrong. You may be able to convince me.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293/17

FRidh commented:
I think at this point in Nixpkgs that very large changes, while desirable, are too much effort, not realistic anymore and thus won't happen. As an example, structured attributes is something that has been worked at for a long time and it hasn't been merged. Because of this, I think we should accept the smaller improvements, *even* when they may not be in line with some bigger picture. I don't like it, but until there is structural funding for people to make large scale improvements, those changes seem to me mostly wishful thinking.

lilyball commented:
@edolstra Given the feedback, are you still opposed to this PR? I would really like to avoid making the perfect stand in the way of the good, and from a user perspective this PR is purely additive (it does add more code to the implementation, but that code is structured as a `makeOverlayable` wrapper and can be skipped over when trying to understand how `mkDerivation` itself works).

The idea of soft-deprecating usage of `rec { }` has come up repeatedly over time, and `mkDerivation` is a major driver of usage of `rec { }`. If I grep through current nixpkgs master I see 19501 hits on the word `rec` and 9019 on `mkDerivation rec`, so almost half of the instances of `rec { }` in nixpkgs are trivially due to `mkDerivation`. Skimming through the results, most of the rest are due to language-specific derivation builders like `buildRustPackage`, `buildPythonApplication`, etc. and I would expect these can be upgraded to have similar behavior as this PR does for `mkDerivation`.

AndersonTorres commented on nixos/doc/manual/from_md/release-notes/rl-2205.section.xml:
```diff
@@ -25,6 +25,33 @@
+        <para>
+          Additionally, <literal>passthru</literal> can now reference
+          <literal>finalAttrs.public</literal> containing the final
+          package, including attributes such as the output paths and
```
Reading this now, I agree with @lilyball  think `package` or `finalPackage` would be a better name for this attribute.

Synthetica9 replied:
(would also help in functions like `testVersion`)

roberth committed:
stdenv.mkDerivation: public -> finalPackage

Artturin commented:
this will be useful for separating tests and using the final package for derivations which use a common builder such as openssh and crystal https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/networking/openssh/default.nix

```nix
diff --git a/pkgs/tools/networking/openssh/common.nix b/pkgs/tools/networking/openssh/common.nix
index 4245d06a1cd..6527e2689de 100644
--- a/pkgs/tools/networking/openssh/common.nix
+++ b/pkgs/tools/networking/openssh/common.nix
@@ -31,7 +31,7 @@
 }:
 
 with lib;
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: rec {
   inherit pname version src;
 
   patches = [
@@ -95,7 +95,7 @@ stdenv.mkDerivation rec {
 
   hardeningEnable = [ "pie" ];
 
-  doCheck = true;
+  doCheck = false;
   enableParallelChecking = false;
   checkInputs = optional (!stdenv.isDarwin) hostname;
   preCheck = lib.optionalString (stdenv.hostPlatform == stdenv.buildPlatform) ''
@@ -162,6 +162,8 @@ stdenv.mkDerivation rec {
   ];
 
   passthru.tests = {
+    testSuite = finalAttrs.finalPackage.overrideAttrs (_: { doCheck = true; });
+
     borgbackup-integration = nixosTests.borgbackup;
   };
 
@@ -174,4 +176,4 @@ stdenv.mkDerivation rec {
     maintainers = (extraMeta.maintainers or []) ++ (with maintainers; [ eelco aneeshusa ]);
     mainProgram = "ssh";
   } // extraMeta;
-}
+})
```

AndersonTorres reviewed (approved):
LGTM

Artturin commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -70,6 +134,7 @@ in
 
 , # TODO(@Ericson2314): Make always true and remove
   strictDeps ? stdenv.hostPlatform != stdenv.buildPlatform
+
```
there's a merge conflict caused by this

patch to fix
```patch
diff --cc pkgs/stdenv/generic/make-derivation.nix
index d1b93874a25,750e878e1ff..00000000000
--- a/pkgs/stdenv/generic/make-derivation.nix
+++ b/pkgs/stdenv/generic/make-derivation.nix
@@@ -69,7 -129,8 +129,8 @@@ makeOverlayable (overrideAttrs
  , doInstallCheck ? config.doCheckByDefault or false
  
  , # TODO(@Ericson2314): Make always true and remove
 -  strictDeps ? stdenv.hostPlatform != stdenv.buildPlatform
 +  strictDeps ? if config.strictDepsByDefault then true else stdenv.hostPlatform != stdenv.buildPlatform
+ 
  , meta ? {}
  , passthru ? {}
  , pos ? # position used in error messages and for meta.position
```

Artturin commented:
This is clearly wanted by many people and needed in nixpkgs. we will probably be able to simplify various language frameworks.

Solving this in nix itself will require lots of work and there are more important things to do

Synthetica9 commented:
Being able to do the following is super cool, can we get that documented?

```nix
hello.overrideAttrs (old: { version = "2.11"; src = old.src.overrideAttrs (old: { outputHash = "sha256-jJzgVy08RO0GcOsc3pgFhOA4tvYsJf396O8SjeFQBL0="; }); })
```

xaverdh commented:
First of all thanks for opening the door to make overriding `src` a little bit more sane. Am I correct that the `mkDerivation` from `pkgs` (*not* `stdenv.mkDerivation`), does not yet work as in this pr?

xaverdh commented:
Ok, sorry its not in the final `pkgs` but it is accessible in the `pkgs` fixed point e.g. [here](https://github.com/NixOS/nixpkgs/blob/64db84dbb05b173d2cfb242e201a1e932325b409/pkgs/tools/filesystems/android-file-transfer/default.nix#L3). I think it was introduced due to Qt issues.

collares commented:
I believe @xaverdh is talking about https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/qt-5/mkDerivation.nix, which `pkgs.libsForQt5.callPackage` automatically injects when called.

Artturin commented:
yeah doesn't work there

```

nix-repl> libsForQt5.mkDerivation (finalAttrs: { name = "hi"; version = finalAttrs.name; } )
error: value is a function while a set was expected

nix-repl> stdenv.mkDerivation (finalAttrs: { name = "hi"; version = finalAttrs.name; } )
«derivation /nix/store/qbr9vsxv0p5lhhf26kbm5639gwyvmr32-hi.drv»
```

Artturin commented on pkgs/applications/misc/hello/default.nix:
```diff
@@ -1,17 +1,18 @@
 }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
```
https://github.com/NixOS/nixpkgs/pull/119942/files#diff-c032817f6a62e23767f4c7706e35eef36e8353aa5ba5d6f9bce8c6a8cc0db01dR53

>  If only a one-argument function is written, the argument has the meaning of `previousAttrs`.

so this should be 
`finalAttrs: previousAttrs:`
right?

Artturin replied:
ah nevermind thats for `overrideAttrs` :sweat_smile:

jtojnar replied:
That is what `overrideAttrs` takes, `stdenv.mkDerivation` is the ground floor so there are no `previousAttrs`.

Ericson2314 commented:
I talked to @roberth about this -- I think this approach is too much of a sledge hammer. Allowing *any* input to `mkDerivation` to depend on the *output* of `mkDerivation` is very unstructured. The need seems to only be for tests/passthru to depend on the output, so I would like to narrow access to the output in that way.

There is a serious cost to creating more and fixed points and cycles every time we need something to depend on something else, and so I would urge caution. Really, we need to have a large conversation about refactoring Nixpkgs to be less rediculous, and stuff like this is but a bandaid in lieu of that conversation.

tjni commented:
I intend to use this to address https://github.com/NixOS/nixpkgs/issues/86349, which needs at least some of this generality. Is there an alternative?

FRidh commented:
> There is a serious cost to creating more and fixed points and cycles every time we need something to depend on something else, and so I would urge caution.

@Ericson2314 do you think there is a substantial increase by using the cycle here? @roberth stats don't show any significant increase, about 0.5%. Is something missing there, maybe because the feature is not user throughout yet? It adds up, of course.

Just one remark I have about merging this now. I am in favor of the feature but I am not so sure about the timing. We're not too far away from branch-off, and while not a breaking change, this change has a large impact as many will want to adopt it, which can also be seen in the PR's being opened referring to this PR. Looking back, it is probably better to state when the PR is being merged (as was originally done) instead of just going ahead with merging.

roberth commented:
@Ericson2314 

> Allowing _any_ input to `mkDerivation` to depend on the _output_ of `mkDerivation` is very unstructured.

Otoh, this keeps the number of scopes lower.

> Really, we need to have a large conversation about refactoring Nixpkgs to be less rediculous, and stuff like this is but a bandaid in lieu of that conversation.

Sounds good. What format would you suggest?

@FRidh 

> do you think there is a substantial increase by using the cycle here?

By adding the `finalAttrs` parameter in the "base" `mkDerivation` call, there's only the extra overhead of an extra `let` binding and an extra function application. This is insignificant and barely worth optimizing in the first place. I'd be surprised if the cost of using `finalAttrs` everywhere would exceed 1%.

> this change has a large impact as many will want to adopt it,

Normally this isn't a consideration, but I agree that it shouldn't affect reviewers especially, yet it does.
I'll put a reminder in the PR description.

# :rotating_light: Please prioritize reviews and fixes for the release!

edolstra commented:
> stats don't show any significant increase, about 0.5%.

Evaluation time bloat happens 0.5% at a time :-) As you say, all these little costs do add up.

lilyball commented:
@Ericson2314

> The need seems to only be for tests/passthru to depend on the output, so I would like to narrow access to the output in that way.

An early motivating example was allowing `src` to depend on the output version, such that overriding the version also updates `src` (this requires overriding `src` as well to fix the `outputHash` but that's not hard). Restricting this feature to just `tests`/`passthru` is a significant restriction that would throw away most of the value of this change.

Ericson2314 commented:
@roberth 

> @Ericson2314 do you think there is a substantial increase by using the cycle here? @roberth stats don't show any significant increase, about 0.5%. Is something missing there, maybe because the feature is not user throughout yet? It adds up, of course.

Oh I just mean conceptual cost. We inch towards a situation like NixOS modules where everything can depend on everything else. This is bit an *inch* and so not that worrisome in itself, but we've only added more cycles never removed them so far, so I am nervous :).

@roberth has done a great job of bench-marking and optimizing, so I am not worried so worried about that.

> Sounds good. What format would you suggest?

I put in #172008 since it is definitely out of scope for now :)

@lilyball

What do you mean "*output* version"? If we have a `version` attribute can't `src` depend on it in the normal `self: super: ...` way? I would invert the dataflow and put the `version` on the `src` instead, and do `inherit (self.src) version;`

lilyball commented:
@Ericson2314 Perhaps I misunderstood you. Were you referring specifically to access to `finalPackage` from anywhere? I do agree that there's a less obvious need to get at `finalPackage` specifically from most places in the derivation, although I haven't thought much on that particular point.

pwaller commented:
I hope this is an appropriate place for this message, apologies if not, but I bisected a failure to a commit which landed in this PR @roberth:

```
nix eval nixpkgs#pkgsStatic.hello
error: value is a function while a set was expected
```

The commit at which it becomes broken is https://github.com/NixOS/nixpkgs/pull/119942/commits/2afc03a084cf736cc76f629a45a720c3c44eb27d

This example appears in various docs which describe how to use pkgsStatic.

As an aside, as a new user of nixos-22.05 (nix v2.8) I can't see an easy way to comprehend where the error is being generated. nix eval doesn't appear to have a --trace option. A naive reading of the error message appears to suggest that 'hello' itself is a function where a set was expected, but I don't believe that is the case.

roberth commented:
The problem seems to be that `pkgsStatic` replaces the `stdenv.mkDerivation` function by an incompatible one using `pkgs/stdenv/adapters.nix`.

> nix eval doesn't appear to have a --trace option

Correct. It's `--show-trace`.

pwaller commented:
> Correct. It's `--show-trace`.

Thanks, confusingly this doesn't appear in the manpage, I knew something like it existed, but couldn't find it in the obvious way?

roberth commented:
@pwaller https://github.com/NixOS/nixpkgs/pull/176385

pwaller commented:
> @pwaller #176385

Thanks, confirmed it evals for me, haven't tried building it since that builds gcc which I don't want to do at the moment for lack of disk space.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/should-our-qt-mkderivation-be-deprecated/20268/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/why-this-override-is-not-working/21530/4

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/overrideattrs-not-taking-effect-how-to-override-fetchfromgithub-and-more/32298/1
